doc: Getting secure TF-M output on nRF5340 DK's (#8942)
nrfconnect/sdk-nrf: doc/nrf/ug_tfm.rst

TF-M employs two UART interfaces for logging: one for the secure part (MCUboot and TF-M), and one for the non-secure application.
The logs arrive on different COM ports on the host PC.

VCom Ports nRF5340 DK
=====================

* v1.0.0: You must connect specific wires on the kit to receive secure logs on the host PC.
  Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector respectively to **RxD** and **TxD**  of the **P24** connector.
  See :ref:`logging_cpunet` on the Working with nRF5340 DK page for more information.
* v2.0.0: There are only two virtual COM ports available on that version. 
  By default one will be used by the non-secure UART peripheral of of the application and the other one by the UART1 peripheral from the network core.
  There are several options to get UART output from the secure TF-M.

> 1. Disable the output for the network core.

Reviewer: Also, does this have to be a numbered list? Are they just the three options? Because now they look more like steps that need to be followed in this order.

> * v1.0.0: You must connect specific wires on the kit to receive secure logs on the host PC.
>   Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector respectively to **RxD** and **TxD**  of the **P24** connector.
>   See :ref:`logging_cpunet` on the Working with nRF5340 DK page for more information.
> * v2.0.0: There are only two virtual COM ports available on that version. 
>   By default one will be used by the non-secure UART peripheral of of the application and the other one by the UART1 peripheral from the network core.
>   There are several options to get UART output from the secure TF-M.
> 1. Disable the output for the network core.
>    For NCS samples see :ref:`ug_multi_image_variables`
>    Then the RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) Pins of TF-M secure can be set to **P1.00** (32) and **P1.01** (33)
> 2. The secure and non-secure UART peripheral can be wired to the same pins.
>    Specifically, wire the pins **P0.25** and **P0.26** to **P0.20** and **P0.22**, respectively.
> 3. If non-secure application, network core output and secure TF-M are all needed simultaneous, then additional UART<->USB hardware is needed, if available a second nRF DK can be used.
>    Then **P0.25** needs to be wired to the TXD pin and **P0.26** to the RXD pin of the external hardware. 
>    These pins will provide the secure TF-M output while the two native COM ports of the DK will be used for the non-secure application and the network core output.

Reviewer: ```suggestion
On the nRF5340 DK v1.0.0, you must connect specific wires on the kit to receive secure logs on the host PC.
Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector respectively to **RxD** and **TxD**  of the **P24** connector.
See :ref:`logging_cpunet` on the Working with nRF5340 DK page for more information.

On the nRF5340 DK v2.0.0, there are only two virtual COM ports available.
By default, one of the ports is used by the non-secure UART peripheral from the application and the other one by the UART1 peripheral from the network core.

There are several options to get UART output from the secure TF-M.

1. Disable the output for the network core.
   For NCS samples, see :ref:`ug_multi_image_variables`.
   The RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) pins of TF-M secure can be set to **P1.00** (32) and **P1.01** (33).

#. The secure and non-secure UART peripheral can be wired to the same pins.
   Specifically, wire the pins **P0.25** and **P0.26** to **P0.20** and **P0.22**, respectively.

#. If non-secure application, network core output and secure TF-M are all needed simultaneously, then additional UART <-> USB hardware is needed.
   A second nRF DK can be used if available.
   Pin **P0.25** needs to be wired to the TXD pin, and **P0.26** to the RXD pin of the external hardware.
   These pins will provide the secure TF-M output, while the two native COM ports of the DK will be used for the non-secure application and the network core output.
```

TF-M employs two UART interfaces for logging: one for the secure part (MCUboot and TF-M), and one for the non-secure application.
The logs arrive on different COM ports on the host PC.

> Virtual COM ports in nRF5340 DK

Reviewer 2: ```suggestion
Virtual COM ports on the nRF5340 DK
```

and then update the ==== (otherwise tooling complains)

> Virtual COM ports in nRF5340 DK
> ===============================

Reviewer 2: ```suggestion
Virtual COM ports on the nRF5340 DK
=============================
```

On the nRF5340 DK v1.0.0, you must connect specific wires on the kit to receive secure logs on the host PC.

> Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector respectively to **RxD** and **TxD**  of the **P24** connector.

Reviewer 2: ```suggestion
Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector to **RxD** and **TxD** respectively of the **P24** connector.
```

> Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector respectively to **RxD** and **TxD** respectively of the **P24** connector.

Reviewer: ```suggestion
Specifically, wire the pins **P0.25** and **P0.26** of the **P2** connector to **RxD** and **TxD** of the **P24** connector respectively.
```

See :ref:`logging_cpunet` on the Working with nRF5340 DK page for more information.

On the nRF5340 DK v2.0.0, there are only two virtual COM ports available.
By default, one of the ports is used by the non-secure UART0 peripheral from the application and the other by the UART1 peripheral from the network core.

> There are several options to get UART output from the secure TF-M.

Reviewer: ```suggestion
There are several options to get UART output from the secure TF-M:
```

> * Disable the output for the network core.

Reviewer 2: ```suggestion
* Disable the output for the network core and change the pins used by TF-M.
```

>   For NCS samples, see :ref:`ug_multi_image_variables`.

Reviewer 2: Do they need to configure a TF-M option that we have no Kconfig for?

If so they would need to be shown these docs, no?

https://docs.zephyrproject.org/3.0.0/guides/tfm/build.html#custom-cmake-arguments

Author: Well again depends on what is running on the network core, but should be realted to TF-M or its config options, but if you build the network core as child image then this points you how to set Kconfigs for a (network core)child image.

Reviewer 2: Why are we linking to ug_multi_image_variables.

The CONFIG_TFM_UART1_RXD_PIN option is not a multi-image variable is it?

Reviewer 2: It's not clear to the reader why he should read "ug_multi_image_variables" is it for changing TF-M pins or disabling network core logging?

Maybe something like:

* Disable the output for the network core and change the pins used by TF-M.
  The network core will usually have an NCS child image.
  To configure a child image see (replace with ref) https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/ug_nrf5340.rst#configuration-of-the-child-image.
  To configure logging in an NCS image see https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_logging.html#logging-in-ncs for the network.
  To change the pins used by TF-M the RXD (:kconfig:option:`CONFIG_TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`CONFIG_TFM_UART1_TXD_PIN`) pins of TF-M can be set to **P1.00** (32) and **P1.01** (33) **by Kconfig options in the application image**.

Reviewer: > It's not clear to the reader why he should read "ug_multi_image_variables" is it for changing TF-M pins or disabling network core logging?
> 
> Maybe something like:
> 
> * Disable the output for the network core and change the pins used by TF-M.
>   The network core will usually have an NCS child image.
>   To configure a child image see (replace with ref) https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/ug_nrf5340.rst#configuration-of-the-child-image.
>   To configure logging in an NCS image see (change with ref) https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_logging.html#logging-in-ncs.
>   To change the pins used by TF-M the RXD (:kconfig:option:`CONFIG_TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`CONFIG_TFM_UART1_TXD_PIN`) Kconfig options in the application image can be set to **P1.00** (32) and **P1.01** (33).
-------
* Disable the output for the network core and change the pins used by TF-M.
  The network core will usually have an NCS child image.
  To configure a child image, see Configuration of the child image section described in :ref:`ug_nrf5340_multi_image`.
  To configure logging in an NCS image, see :ref:`ug_logging`.
  To change the pins used by TF-M, the RXD (:kconfig:option:`CONFIG_TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`CONFIG_TFM_UART1_TXD_PIN`) Kconfig options in the application image can be set to **P1.00** (32) and **P1.01** (33).

Another solution for child image config is to add label for the Configuration of the child image section, and reference it directly. Add e.g.
.. _ug_nrf5340_multi_image_child_image_config:

Configuration of the child image
================================

>   The RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) pins of TF-M secure can be set to **P1.00** (32) and **P1.01** (33).

Reviewer 2: Have you followed these intructions and can confirm they work?

Reviewer 2: The instructions seem to only cover how to configure the secure pins, is there anything we can do to help them understand how to disable output on the network core?

Author: > Have you followed these instructions and can confirm they work?

Sure tests it on my desk, but happy if somebody wants to test this as well.

Author: > The instructions seem to only cover how to configure the secure pins, is there anything we can do to help them understand how to disable output on the network core?

Well that depends on what you run on the network core, that will get a bit out of scope here in my opinion

>   The RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) pins of TF-M secure can be set to **P1.00** (32) and **P1.01** (33).

Reviewer 2: ```suggestion
  The RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) pins of TF-M secure can be set to **P1.00** (32) and **P1.01** (33).
```

It's a Kconfig option, but it's prefixed with TFM_ ?

Author: I just used the define, name but realised that we use as well CONFIG_ for all the kconfig options as well, changed it

>   The RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) pins of TF-M secure can be set to **P1.00** (32) and **P1.01** (33).

Reviewer 2: ```suggestion
  The RXD (:kconfig:option:`TFM_UART1_RXD_PIN`) and TXD (:kconfig:option:`TFM_UART1_TXD_PIN`) pins of TF-M can be set to **P1.00** (32) and **P1.01** (33).
```

> * The secure and non-secure UART peripheral can be wired to the same pins.

Reviewer 2: ```suggestion
* The secure and non-secure UART peripherals can be wired to the same pins.
```

>   Specifically, wire the pins **P0.25** and **P0.26** to **P0.20** and **P0.22**, respectively.

Reviewer 2: ```suggestion
  Specifically, physically wire together the pins **P0.25** and **P0.26** to **P0.20** and **P0.22**, respectively.
```

> * If non-secure application, network core output and secure TF-M are all needed simultaneously, then additional UART <-> USB hardware is needed.

Reviewer 2: ```suggestion
* If the non-secure application, network core and TF-M image are all needed simultaneously, then additional UART <-> USB hardware is needed.
```

Author: Changed it to `If the non-secure application, network core and TF-M outputs are all needed simultaneously` this is still about the output not the images itself

TF-M employs two UART interfaces for logging: one for the secure part (MCUboot and TF-M), and one for the non-secure application.
The logs arrive on different COM ports on the host PC.

Virtual COM ports on the nRF5340 DK
===================================

> On the nRF5340 DK v1.0.0, you must connect specific wires on the kit to receive secure logs on the host PC.

Reviewer: Nice if we could add space (empty line) before the first sentence, otherwise I think it looks good :)
